src/org/javarosa/xform/parse/XmlTextConsolidator.java

```diff
@@ -0,0 +1,137 @@
+package org.javarosa.xform.parse;
+
+import org.javarosa.core.util.CacheTable;
+import org.kxml2.kdom.Element;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Stack;
+
+import static org.kxml2.kdom.Node.ELEMENT;
+import static org.kxml2.kdom.Node.TEXT;
+
+class XmlTextConsolidator {
+
+    private static final int NUM_DELETIONS_PERFORMANCE_THRESHOLD = 100;
+    private static final int NUM_CHILDREN_PERFORMANCE_THRESHOLD = 100;
+
+    /**
+     * According to Clayton Sims, in Feb 2012:
+     * For escaped unicode strings we end up with a lot of cruft, so we really
+     * want to go through and convert the kxml parsed text (which have lots of
+     * characters each as their own string) into one single string
+     */
+    static void consolidateText(CacheTable<String> stringCache, Element rootElement) {
+        Stack<Element> q = new Stack<>();
+        q.push(rootElement);
+        while (!q.isEmpty()) {
+            final Element e = q.pop();
+            final Set<Integer> removeIndexes = new HashSet<>();
+            String accumulator = "";
+            for (int i = 0; i < e.getChildCount(); ++i) {
+                final int type = e.getType(i);
+                if (type == TEXT) {
+                    accumulator += e.getText(i);
+                    removeIndexes.add(i);
+                } else {
+                    if (type == ELEMENT) {
+                        q.add(e.getElement(i));
+                    }
+                    if (nonBlank(accumulator)) {
+                        e.addChild(i++, TEXT, maybeInternedString(stringCache, accumulator));
+                    }
+                    accumulator = "";
+                }
+            }
+            if (nonBlank(accumulator)) {
+                e.addChild(TEXT, maybeInternedString(stringCache, accumulator));
+            }
+            if (highVolumeApproachNeeded(removeIndexes.size(), e.getChildCount())) {
+                rebuildChildrenAndAttributes(e, removeIndexes);
+            } else {
+                removeDeletedChildren(e, removeIndexes);
+            }
+        }
+    }
+
+    private static boolean nonBlank(String accumulate) {
+        return !accumulate.trim().isEmpty();
+    }
+
+    private static String maybeInternedString(CacheTable<String> stringCache, String accumulate) {
+        return stringCache == null ? accumulate : stringCache.intern(accumulate);
+    }
+
+    /**
+     * Determines whether to use the simpler, much slower, child deletion method, or the more complicated
+     * one, which performs well even with tens of thousands of deletions or children.
+     *
+     * @param numDeletions the number of deletions to be done
+     * @param numChildren  the number of children
+     * @return whether to use the high volume approach
+     */
+    private static boolean highVolumeApproachNeeded(int numDeletions, int numChildren) {
```

Comment:
Why do we want to support both child deletion approaches? It seems like we can keep just `rebuildChildrenAndAttributes` if it's *always* faster than `removeDeletedChildren`. On the other hand, we may want to keep it as an explanation why we need the complex `rebuildChildrenAndAttributes`.

Reply:
The new approach performs better only for forms with many children.